File: mvp/test-fixtures/python/01-fastapi-backend/src/models/Product.py
from typing import Optional
from datetime import datetime


class Product:
    """Product model representing an item in the catalog"""
    
    id: int
    name: str
    description: str
    price: float
    stock: int
    created_at: datetime
    
    def __init__(self, id: int, name: str, price: float, description: str = "", stock: int = 0):
        self.id = id
        self.name = name
        self.description = description
        self.price = price
        self.stock = stock
        self.created_at = datetime.now()
    
    def to_dict(self) -> dict:
        """Convert product to dictionary"""
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'price': self.price,
            'stock': self.stock,
            'created_at': self.created_at.isoformat()
        }
    
    def update_stock(self, quantity: int) -> None:
        """Update product stock"""
        self.stock += quantity
    
    def is_available(self) -> bool:
        """Check if product is in stock"""
        return self.stock > 0
    
    def apply_discount(self, percentage: float) -> float:
        """Apply discount and return new price"""
        discount = self.price * (percentage / 100)
        return self.price - discount


class ProductRepository:
    """Repository for Product data access"""
    
    def __init__(self):
        self.products = {}
        self.next_id = 1
    
    def find_by_id(self, id: int) -> Optional[Product]:
        """Find product by ID"""
        return self.products.get(id)
    
    def find_all(self) -> list:
        """Get all products"""
        return list(self.products.values())
    
    def save(self, product: Product) -> Product:
        """Save product to repository"""
        if product.id == 0:
            product.id = self.next_id
            self.next_id += 1
        self.products[product.id] = product
        return product
    
    def delete(self, id: int) -> bool:
        """Delete product by ID"""
        if id in self.products:
            del self.products[id]
            return True
        return False

File: test-fixtures/php/01-laravel-backend/src/Models/Product.php
<?php

namespace App\Models;

use DateTime;

/**
 * Product model representing an item in the catalog
 */
class Product
{
    private int $id;
    private string $name;
    private string $description;
    private float $price;
    private int $stock;
    private DateTime $createdAt;
    
    public function __construct(
        int $id,
        string $name,
        float $price,
        string $description = "",
        int $stock = 0
    ) {
        $this->id = $id;
        $this->name = $name;
        $this->description = $description;
        $this->price = $price;
        $this->stock = $stock;
        $this->createdAt = new DateTime();
    }
    
    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'description' => $this->description,
            'price' => $this->price,
            'stock' => $this->stock,
            'created_at' => $this->createdAt->format('Y-m-d H:i:s')
        ];
    }
    
    public function updateStock(int $quantity): void
    {
        $this->stock += $quantity;
    }
    
    public function isAvailable(): bool
    {
        return $this->stock > 0;
    }
    
    public function applyDiscount(float $percentage): float
    {
        $discount = $this->price * ($percentage / 100);
        return $this->price - $discount;
    }
    
    public function getId(): int
    {
        return $this->id;
    }
    
    public function getName(): string
    {
        return $this->name;
    }
    
    public function getPrice(): float
    {
        return $this->price;
    }
}

/**
 * Repository for Product data access
 */
class ProductRepository
{
    private array $products = [];
    private int $nextId = 1;
    
    public function findById(int $id): ?Product
    {
        return $this->products[$id] ?? null;
    }
    
    public function findAll(): array
    {
        return array_values($this->products);
    }
    
    public function save(Product $product): Product
    {
        if ($product->getId() === 0) {
            $this->nextId++;
        }
        $this->products[$product->getId()] = $product;
        return $product;
    }
    
    public function delete(int $id): bool
    {
        if (isset($this->products[$id])) {
            unset($this->products[$id]);
            return true;
        }
        return false;
    }
}
